src/app/content/components/Page/highlightManager.ts

```diff
@@ -163,10 +174,37 @@ export default (container: HTMLElement, getProp: () => HighlightProp) => {
         ;
 
       highlighter.clearFocus();
-      const focusedId = getProp().focused;
-      const focused = focusedId && highlighter.getHighlight(focusedId);
+      const { scrollTarget, focus, focused: focusedId, highlightsLoaded, loggedOut, page } = getProp();
+      const focused = focusedId ? highlighter.getHighlight(focusedId) : null;
+      const stateEstablished = (highlightsLoaded || (loggedOut && page));
+
+      const highlightScrollTarget = scrollTarget && isHighlightScrollTarget(scrollTarget) ? scrollTarget : null;
+      const scrollTargetHighlight = highlightScrollTarget && highlighter.getHighlight(highlightScrollTarget.id);
+
+      if (options && stateEstablished) {
+        if (scrollTargetHighlight) {
+          options.clearError();
+        } else if (highlightScrollTarget) {
+          options.setError(highlightScrollTarget.id);
+        }
+      }
+
+      let toFocus: Highlight | null = null;
       if (focused) {
-        focused.focus();
+        toFocus = focused;
+      } else if (
+        !pendingHighlight
+        && !prevProps.focused
+        && scrollTargetHighlight
+      ) {
+        toFocus = scrollTargetHighlight;
+      }
+
+      if (toFocus) {
+        toFocus.focus();
+        if (toFocus.id !== focusedId) {
+          focus(toFocus.id);
```

Comment:
would it be simpler to have a redux hook watching for `locationChange` containing a scrolltarget and dispatch focus from there? i feel like there is a lot of stuff going on right in here and moving that out might reduce a dimension of it

Reply:
It would have to be triggered after `highlightsLoaded` is truthy and it would split focusing logic into one more place which can be tricky to debug in the future and we would only get rid of:
```js
} else if (
  !pendingHighlight
  && !prevProps.focused
  && scrollTargetHighlight
) {
  toFocus = scrollTargetHighlight;
}
```
and
```js
if (toFocus.id !== focusedId) {
  focus(toFocus.id);
}
```
I think I can just create a helper like `const toFocus = getHighlightToFocus(...)`